src/manual/devel/best-practices/microservices.md

```diff
@@ -0,0 +1,307 @@
+Its README describes how to set it up generally, but for a TaskCluster service, you will need to install `babel-compile` and `babel-preset-taskcluster`.
+Include the following in your `package.json`:
+
+```js
+  "scripts": {
+    "compile": "babel-compile -p taskcluster src:lib test:.test",
+    "pretest": "npm run compile",
+    "install": "npm run compile",
+  }
+```
+
+### Dependencies
+
+Try to keep up-to-date with the latest versions of all TaskCluster libraries.
+In general, the implications of updating these libraries should be clear, and the authors are easy to find when things go badly.
+
+Other dependencies should also be kept up-to-date as much as possible.
+Tools like [Greenkeeper](https://greenkeeper.io/) can be very useful for this purpose.
+
+### Shrinkwrap
+
+Dependencies for deployed services are handled with [`npm shrinkwrap`](https://docs.npmjs.com/cli/shrinkwrap).
+This is to ensure that we are deploying a consistent tree of package versions, and that dependency upgrades are controlled and not automatic or surprising.
+
+When creating a new service, run `npm shrinkwrap` to generate an `npm-shrinkwrap.json` based on the packages you have installed in your development environment, then check that file into the Git repository.
+When beginning development on an existing service, the usual `npm install` will install the specific versions named in the `npm-shrinkwrap.json`.
+
+*Important*: Do not use `devDependencies`, as npm versions before 3.0 do not deal with them correctly and will cause failures in deployment that do not appear in local runs or Travis.
+Instead, list all dependencies in `dependencies`.
+
+When changing a service's dependencies, the easiest approach is to use `npm install --save` to install the package and update `package.json` at the same time.
+Once this is complete, run `npm shrinkwrap` and add the modified `npm-shrinkwrap.json` to your commit.
+Pull requests with a modified `package.json` that do not have a corresponding `npm-shrinkwrap.json` diff should be viewed with suspicion!
+
+If is the service owner's responsibility to keep dependencies up to date.
+The `npm outdated` command gives a useful overview of available updates.
+In general, try to keep packages up to date within semver constraints (so, fix things displayed in red in `npm outdated`), but be cautious that the new code you are incorporating into your service is trustworthy.
+In an ideal world, that means a thorough security review.
+In the real world, that probably means a lot less.
+
+## Testing
+
+### Test Setup
+
+Use Mocha to run unit tests, in the `test/` directory.
+Include the following in `mocha.opts`, most critically the `--ui tdd`.
+
+```
+--ui tdd
+--timeout 30s
+--reporter spec
+```
+
+Name the test files `test/*_test.js`, so they will be matched by the `npm test` script given above.
+These files should require production code from `../lib/`: `foo = require('../lib/foo')`.
```

Comment:
```
(v5.12.0) dustin@lamport ~/p/taskcluster-docs [microservices-best-practices*] $ ls -d1 ../{taskcluster-,aws-}*/.test
ls: cannot access '../aws-*/.test': No such file or directory
../taskcluster-auth/.test
../taskcluster-github/.test
../taskcluster-hooks/.test
../taskcluster-host-secrets/.test
../taskcluster-lib-api/.test
../taskcluster-lib-app/.test
../taskcluster-lib-loader/.test
../taskcluster-lib-monitor/.test
../taskcluster-lib-scopes/.test
../taskcluster-lib-testing/.test
../taskcluster-lib-validate/.test
../taskcluster-login/.test
../taskcluster-mercurial/.test
../taskcluster-pulse/.test
../taskcluster-purge-cache/.test
../taskcluster-queue/.test
../taskcluster-stats-collector/.test
(v5.12.0) dustin@lamport ~/p/taskcluster-docs [microservices-best-practices*] $
(v5.12.0) dustin@lamport ~/p/taskcluster-docs [microservices-best-practices*] $ ls -d1 ../{taskcluster-,aws-}*/test-src
ls: cannot access '../taskcluster-*/test-src': No such file or directory
../aws-provisioner/test-src
```
so `.test` seems to be the current consensus.  I'd like to get this doc landed with current practices, but certainly PRs are awesome afterward to modify it (especially if that comes along with actually modifying all or most of the services!).  I think Jonas has a few in mind, too :)